### PR TITLE
feat: add superchain erc20 bridge

### PR DESCRIPTION
**Pull request by agusduha**
Closes OPT-224

### CODE REVIEW

File: packages/contracts-bedrock/snapshots/storageLayout/SuperchainERC20Bridge.json [hunk 1]
```diff
@@ -0,0 +1 @@
+[]
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/snapshots/storageLayout/SuperchainERC20Bridge.json:1]: Are These empty brackets really needed?

File: packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    /// @notice Semantic version.
+    /// @custom:semver 1.0.0-beta.1
+    string public constant version = "1.0.0-beta.1";
+
+    /// @notice Sends tokens to some target address on another chain.
+    /// @param _token   Token to send.
+    /// @param _to      Address to send tokens to.
+    /// @param _amount  Amount of tokens to send.
+    /// @param _chainId Chain ID of the destination chain.
+    function sendERC20(address _token, address _to, uint256 _amount, uint256 _chainId) external override {
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol]: No need to explicitly override an interface
```suggestion
    function sendERC20(address _token, address _to, uint256 _amount, uint256 _chainId) external {
```

File: packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+        emit SendERC20(address(_token), msg.sender, _to, _amount, _chainId);
+    }
+
+    /// @notice Relays tokens received from another chain.
+    /// @param _token  Token to relay.
+    /// @param _from   Address of the msg.sender of sendERC20 on the source chain.
+    /// @param _to     Address to relay tokens to.
+    /// @param _amount Amount of tokens to relay.
+    function relayERC20(address _token, address _from, address _to, uint256 _amount) external override {
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol]: Same here
```suggestion
    function relayERC20(address _token, address _from, address _to, uint256 _amount) external {
```

File: packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+        if (IL2ToL2CrossDomainMessenger(MESSENGER).crossDomainMessageSender() != address(this)) {
+            revert InvalidCrossDomainSender();
+        }
+
+        uint256 source = IL2ToL2CrossDomainMessenger(MESSENGER).crossDomainMessageSource();
+
+        IMintableAndBurnableERC20(_token).mint(_to, _amount);
+
+        emit RelayERC20(address(_token), _from, _to, _amount, source);
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol]:
```suggestion
        emit RelayERC20(_token, _from, _to, _amount, source);
```

File: packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    /// @param _to      Address to send tokens to.
+    /// @param _amount  Amount of tokens to send.
+    /// @param _chainId Chain ID of the destination chain.
+    function sendERC20(address _token, address _to, uint256 _amount, uint256 _chainId) external override {
+        IMintableAndBurnableERC20(_token).burn(msg.sender, _amount);
+
+        bytes memory message = abi.encodeCall(this.relayERC20, (_token, msg.sender, _to, _amount));
+        IL2ToL2CrossDomainMessenger(MESSENGER).sendMessage(_chainId, address(this), message);
+
+        emit SendERC20(address(_token), msg.sender, _to, _amount, _chainId);
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol]:
```suggestion
        emit SendERC20(_token, msg.sender, _to, _amount, _chainId);
```

File: packages/contracts-bedrock/src/L2/interfaces/ISuperchainERC20Bridge.sol [hunk 6]
```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/src/L2/interfaces/ISuperchainERC20Bridge.sol:2]:
```suggestion
pragma solidity =0.8.25;
```

File: packages/contracts-bedrock/src/L2/interfaces/IMintableAndBurnableERC20.sol [hunk 7]
```diff
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/src/L2/interfaces/IMintableAndBurnableERC20.sol:2]: Only if it is gonna be used on the `SuperchainERC20Bridge` contract
```suggestion
pragma solidity =0.8.25;
```

File: packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+/// @notice The SuperchainERC20Bridge allows for the bridging of ERC20 tokens to make them fungible across the
+///         Superchain. It builds on top of the L2ToL2CrossDomainMessenger for both replay protection and domain
+///         binding.
+contract SuperchainERC20Bridge is ISuperchainERC20Bridge, ISemver {
+    /// @notice Address of the L2ToL2CrossDomainMessenger Predeploy.
+    address internal constant MESSENGER = Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER;
+
+    /// @notice Semantic version.
+    /// @custom:semver 1.0.0-beta.1
+    string public constant version = "1.0.0-beta.1";
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol:24]: Question: Should we maintain the same version for the redesign?

**Review by 0xDiscotech — COMMENTED** [review]
Looking good so far sir!
Left few minor comments and a question.

Also, after running `just pre-pr` the semver failed, so it needs fix:
image

File: packages/contracts-bedrock/snapshots/storageLayout/SuperchainERC20Bridge.json (hunk 1 above)
Comment (agusduha) [reply packages/contracts-bedrock/snapshots/storageLayout/SuperchainERC20Bridge.json:1]: This is autogenerated by the snapshot so the pre-pr will add it always

File: packages/contracts-bedrock/src/L2/interfaces/ISuperchainERC20Bridge.sol (hunk 6 above)
Comment (agusduha) [reply packages/contracts-bedrock/src/L2/interfaces/ISuperchainERC20Bridge.sol:2]: We are leaving interfaces pragmas open to facilitate integrations

File: packages/contracts-bedrock/src/L2/interfaces/IMintableAndBurnableERC20.sol (hunk 7 above)
Comment (agusduha) [reply packages/contracts-bedrock/src/L2/interfaces/IMintableAndBurnableERC20.sol:2]: It is also being used in the `L2StandardBridgeInterop` that is `0.8.15` so it needs to be open. Also all interfaces need to be like this to support integrations, at least while there are differents pragmas across the code

File: packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol (hunk 8 above)
Comment (agusduha) [reply packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol:24]: The same version as what? I put this one because it is the first iteration of this bridge contract, but we can change it

File: packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol [hunk 9]
```diff
@@ -16,7 +15,7 @@ import { IL2ToL2CrossDomainMessenger } from "src/L2/interfaces/IL2ToL2CrossDomai
 /// @notice The SuperchainERC20Bridge allows for the bridging of ERC20 tokens to make them fungible across the
 ///         Superchain. It builds on top of the L2ToL2CrossDomainMessenger for both replay protection and domain
 ///         binding.
-contract SuperchainERC20Bridge is ISuperchainERC20Bridge, ISemver {
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol]: Why did you remove the `ISemver`?

Comment (agusduha) [reply packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol]: I moved `ISemver` inside `ISuperchainERC20Bridge.sol` because `check-interface` script was failing due to having incompatible ABI between the contract and its interface (`version` was missing at the interface level)

Comment (agusduha) [reply packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol]: I though it was cleaner to put it like that, but also another `version` func can be added

File: packages/contracts-bedrock/src/L2/L2StandardBridgeInterop.sol [hunk 10]
```diff
@@ -40,9 +40,9 @@ contract L2StandardBridgeInterop is L2StandardBridge {
     event Converted(address indexed from, address indexed to, address indexed caller, uint256 amount);
 
     /// @notice Semantic version.
-    /// @custom:semver +interop
+    /// @custom:semver +interop-beta.1
     function version() public pure override returns (string memory) {
-        return string.concat(super.version(), "+interop");
+        return string.concat(super.version(), "+interop-beta.1");
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/src/L2/L2StandardBridgeInterop.sol:45]: Now I'm not very sure about this. Maybe we can ask the op team how to handle versioning on this cases, when we are updating a contract that was never released under any version, but that differs from the latest one

File: packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol (hunk 9 above)
Comment (0xDiscotech) [inline packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol]: Why are you removing the `ISemver` if the contract has a `version`?

Comment (0xDiscotech) [reply packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol]: perfect then!

**Review by 0xDiscotech — APPROVED** [review]
Nice job!

File: packages/contracts-bedrock/src/L2/L2StandardBridgeInterop.sol (hunk 10 above)
Comment (agusduha) [reply packages/contracts-bedrock/src/L2/L2StandardBridgeInterop.sol:45]: Solved here https://discord.com/channels/1244729134312198194/1244736279384686614/1288553525323890760
